timeskew/timeskew.go

```diff
@@ -0,0 +1,117 @@
+package timeskew
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/beevik/ntp"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	// DefaultServer will be internally resolved to the closest available.
+	// also it rarely queries same server more than once.
+	DefaultServer = "pool.ntp.org"
+
+	// DefaultAttempts defines how many servers we will query
+	DefaultAttempts = 3
+
+	// DefaultUpdatePeriod defines how often time will be queried from ntp.
+	DefaultUpdatePeriod = 2 * time.Minute
+)
+
+type ntpQuery func(string) (*ntp.Response, error)
+
+func computeOffset(queryMethod ntpQuery, server string, attempts int) (time.Duration, error) {
+	offsets := make([]time.Duration, 5)
+	// lowest and highest will be cutoff to prevent reading data from out of sync ntp servers
+	for i := 0; i < attempts+2; i++ {
+		response, err := queryMethod(server)
+		if err != nil {
+			return 0, err
+		}
+		offsets[i] = response.ClockOffset
+	}
+	sort.SliceStable(offsets, func(i, j int) bool {
+		return offsets[i] > offsets[j]
+	})
+	var sum time.Duration
+	for i := 1; i <= attempts; i++ {
+		sum += offsets[i]
+	}
+	return sum / time.Duration(attempts), nil
+}
+
+// NewDefaultTimeSource initializes time source with default config values.
+func NewDefaultTimeSource() *TimeSource {
```

Comment:
Why `Default`? There's no other constructing function and default initialisation seems regular.

Reply:
because it used all defaults declared in the package. i didn't add "normal" constructor because i don't have a use case for it

Reply:
I'd then stick to either `timesource.New()` or `timesource.Default()` as a method name